File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-scala/demo/src/main/scala/bz/stub/parallelconsumer/client/scaladsl/demo/ArmResult.scala
// Copyright (C) 2026 Antony Stubbs and contributors

package bz.stub.parallelconsumer.client.scaladsl.demo

import scala.concurrent.duration.FiniteDuration

/**
 * What one arm achieved: how long it took, over how many records, and across how many distinct keys.
 *
 * ==`processed` and `keys` are the deterministic pair, and that is why they are here==
 *
 * Throughput alone cannot show the work happened - a short arm reports a flattering rate, and a rate
 * is all a reader would have to judge it by. `processed` must equal the target, so an arm that fell
 * short is visibly a failed arm rather than a fast one; `keys` is the distinct keys the arm actually
 * saw, which shows the backlog was spread rather than one key repeated.
 *
 * '''Unlike `elapsed` and the rate, both are the same in every language''' - the same records, the
 * same key space, so the same two figures. That is what makes the two tables comparable across
 * eleven demos when no timing ever could be, and it is what `bin/ci-demo-conformance.sh` leans on.
 *
 * '''There is no latency field, and its absence is the contract rather than an omission.''' The
 * backlog is pre-produced, so the workload is closed-loop and a per-record timing would be flattered
 * by however far an arm had fallen behind. Throughput is the only honest number this shape can
 * produce, so it is the only one there is a place to put.
 */
private[demo] final case class ArmResult(
    arm: String,
    elapsed: FiniteDuration,
    processed: Int,
    keys: Int) {

  def ratePerSecond: Double = {
    val seconds = elapsed.toNanos / 1000000000d
    if (seconds > 0) processed / seconds else 0d
  }
}

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-dotnet/demo/Bz.Stub.ParallelConsumer.Proxy.Client.Demo/ArmResult.cs
// Copyright (C) 2026 Antony Stubbs and contributors

namespace Bz.Stub.ParallelConsumer.Proxy.Client.Demo;

/// <summary>
/// What one arm achieved: how long it took, over how many records, and across how many keys.
/// </summary>
/// <remarks>
/// <see cref="Processed"/> and <see cref="UniqueKeys"/> are the two DETERMINISTIC figures in the
/// tables, and that is why they are here rather than only in a log line. Throughput alone cannot
/// show the work happened - a short arm is a failed arm rather than a fast one - and elapsed and
/// msg/s can never be compared between two languages on two machines, while these two must agree
/// exactly. <c>bin/ci-demo-conformance.sh</c> leans on precisely that.
/// </remarks>
/// <param name="Arm">The arm's label, as it appears in the tables - the role AND the client.</param>
/// <param name="Elapsed">Wall clock, measured from the first poll to the last outcome.</param>
/// <param name="Processed">How many records the arm's own function ran on.</param>
/// <param name="UniqueKeys">
/// How many distinct record keys the arm's own function saw. It shows the backlog was really spread
/// rather than one key repeated, which a record count alone cannot.
/// </param>
internal sealed record ArmResult(string Arm, TimeSpan Elapsed, int Processed, int UniqueKeys)
{
    /// <summary>
    /// Throughput, which is the only TIMING figure this demo reports.
    /// </summary>
    /// <remarks>
    /// The backlog is pre-produced, so the workload is closed-loop and a per-record latency would be
    /// flattered by however far an arm had fallen behind. Throughput is the honest number this shape
    /// can produce, so no arm reports any other timing.
    /// </remarks>
    public double RatePerSecond => Elapsed.TotalSeconds > 0 ? Processed / Elapsed.TotalSeconds : 0;
}
